torch/optim/swa_utils.py

```diff
@@ -0,0 +1,262 @@
+        >>>                                     T_max=300)
+        >>> swa_start = 160
+        >>> swa_scheduler = SWALR(optimizer, swa_lr=0.05)
+        >>> for i in range(300):
+        >>>      for input, target in loader:
+        >>>          optimizer.zero_grad()
+        >>>          loss_fn(model(input), target).backward()
+        >>>          optimizer.step()
+        >>>      if i > swa_start:
+        >>>          swa_model.update_parameters(model)
+        >>>          ema_model.update_parameters(model)
+        >>>          swa_scheduler.step()
+        >>>      else:
+        >>>          scheduler.step()
+        >>>
+        >>> # Update bn statistics for the swa_model at the end
+        >>> torch.optim.swa_utils.update_bn(loader, swa_model) 
+
+    .. note::
+        When using SWA with models containing Batch Normalization you may 
+        need to update the activation statistics for Batch Normalization.
+        You can do so by using :meth:`torch.optim.swa_utils.update_bn` utility.
+
+    .. note::
+        :attr:`avg_fun` is not saved in the :meth:`state_dict` of the model.
+
+    .. note::
+        When :meth:`update_parameters` is called for the first time (i.e. 
+        :attr:`n_averaged` is `0`) the parameters of `model` are copied
+        to the parameters of :class:`AveragedModel`. For every subsequent
+        call of :meth:`update_parameters` the function `avg_fun` is used
+        to update the parameters.
+
+    .. _Averaging Weights Leads to Wider Optima and Better Generalization:
+        https://arxiv.org/abs/1803.05407
+    .. _There Are Many Consistent Explanations of Unlabeled Data: Why You Should
+        Average:
+        https://arxiv.org/abs/1806.05594
+    .. _SWALP: Stochastic Weight Averaging in Low-Precision Training:
+        https://arxiv.org/abs/1904.11943
+    .. _Stochastic Weight Averaging in Parallel: Large-Batch Training That 
+        Generalizes Well:
+        https://arxiv.org/abs/2001.02312
+    """
+    def __init__(self, model, device=None, avg_fun=None):
+        super(AveragedModel, self).__init__()
+        self.module = deepcopy(model)
+        if device is not None:
+            self.module = self.module.to(device)
+        self.register_buffer('n_averaged', torch.tensor(0, dtype=torch.long))
+        if avg_fun is None:
+            def avg_fun(p_avg, p, n_avg): 
+                return p_avg + (p - p_avg) / (n_avg + 1)
+        self.avg_fun = avg_fun
+
+    def forward(self, *args, **kwargs):
+        return self.module(*args, **kwargs)
+
+    def update_parameters(self, model):
+        for p_swa, p_model in zip(self.parameters(), model.parameters()):
+            device = p_swa.device
+            p_model_ = p_model.data.to(device)
+            if self.n_averaged == 0:
+                p_swa.data.copy_(p_model_)
+            else:
+                p_swa.data.copy_(self.avg_fun(p_swa.data, p_model_,
+                                              self.n_averaged))
```

Comment:
@albanD -- If I remember our discussions correctly, `.data` shouldn't be used at all, and `torch.no_grad` should be used instead. Thoughts?